orbit/pkg/table/fleetd_logs/fleetd_logs.go

```diff
@@ -0,0 +1,166 @@
+package fleetd_logs
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/osquery/osquery-go/plugin/table"
+	"github.com/rs/zerolog"
+)
+
+// No timezone, always return in UTC. Use this format because SQLite3
+// knows how to parse it.
+// See https://www.sqlite.org/lang_datefunc.html
+const timeFormatString = "2006-01-02 15:04:05.999999999"
+
+var Logger = logger{}
+var MaxEntries uint = 10_000
+
+func TablePlugin() *table.Plugin {
+	columns := []table.ColumnDefinition{
+		table.TextColumn("time"),
+		table.TextColumn("level"),
+		table.TextColumn("payload"),
+		table.TextColumn("message"),
+		table.TextColumn("error"),
+	}
+
+	return table.NewPlugin("fleetd_logs", columns, generate)
+}
+
+func generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	output := []map[string]string{}
+
+	for _, entry := range Logger.logs {
+		row := make(map[string]string, 5)
+		// It would be nice if we could return NULL instead of an
+		// empty string when the error is empty
+		row["time"] = entry.Time
+		row["level"] = entry.Level.String()
+		row["payload"] = string(entry.Payload)
+		row["message"] = entry.Message
+		row["error"] = entry.Error
+		output = append(output, row)
+	}
+	return output, nil
+}
+
+type Event struct {
+	Time    string
+	Level   zerolog.Level
+	Payload []byte
+	Message string
+	Error   string
+}
+
+type logger struct {
+	writeMutex sync.Mutex
+	logs       []Event
+}
+
+func (l *logger) Write(event []byte) (int, error) {
+	msg, err := processLogEntry(event)
+	if err != nil {
+		return 0, fmt.Errorf("fleet_logs.Write: %w", err)
+	}
+
+	l.writeMutex.Lock()
+	defer l.writeMutex.Unlock()
+
+	l.logs = append(l.logs, msg)
+
+	if MaxEntries > 0 && len(l.logs) > int(MaxEntries) {
+		l.logs = l.logs[len(l.logs)-int(MaxEntries):]
+	}
+
+	return len(event), nil
+}
+
+func (l *logger) WriteLevel(level zerolog.Level, event []byte) (int, error) {
+	msg, err := processLogEntry(event)
+	if err != nil {
+		return 0, fmt.Errorf("fleet_logs.WriteLevel: %w", err)
+	}
+
+	msg.Level = level
+
+	l.writeMutex.Lock()
+	defer l.writeMutex.Unlock()
+
+	l.logs = append(l.logs, msg)
+
+	if MaxEntries > 0 && len(l.logs) > int(MaxEntries) {
+		l.logs = l.logs[len(l.logs)-int(MaxEntries):]
+	}
+
+	return len(event), nil
+}
+
+func processLogEntry(event []byte) (Event, error) {
+	var evt map[string]interface{}
+	dec := json.NewDecoder(bytes.NewReader(event))
+	dec.UseNumber()
+	if err := dec.Decode(&evt); err != nil {
+		return Event{}, fmt.Errorf("cannot decode: %w", err)
+	}
```

Comment:
this is a tricky API, I learned this [from Martin](https://github.com/fleetdm/fleet/pull/7353#discussion_r952595761):

> The problem with json.Decoder is that it consumes a stream of JSON values, so when it's done decoding one value, it is not guaranteed that it consumed the whole input (e.g. https://pkg.go.dev/encoding/json#example-Decoder). What we should do to enforce that is to check that it returns io.EOF (and eat that error as a success)

Reply:
I was thinking just checking for `io.EOF`? to prevent malformed input

Reply:
Malformed input should return a non-EOF error I think? `io.EOF` will only get returned if we read the input in a loop and there is no more data after the first object is parsed

Reply:
not always, take this for example:

`{"foo": "bar"}{`

here decoder will decode the first token, however the JSON is invalid. It's a footgun of the decoder, if you're decoding a single token you should keep consuming tokens until you reach `io.EOF`

in this case, since you're expecting a single token, we can fail if the next one is not `io.EOF`

Reply:
Understood, I have added support for `io.EOF` 🙂

Reply:
(you can see in the official docs I linked an elegant way to do it)